pkg/imgpkg/bundle/locations.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package bundle
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/cppforlife/go-cli-ui/ui"
+	"github.com/google/go-containerregistry/pkg/name"
+	regv1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	"github.com/google/go-containerregistry/pkg/v1/types"
+	"github.com/k14s/imgpkg/pkg/imgpkg/image"
+	"github.com/k14s/imgpkg/pkg/imgpkg/plainimage"
+	"github.com/k14s/imgpkg/pkg/imgpkg/util"
+)
+
+const (
+	locationsTag string = "%s-%s.locations.imgpkg"
+)
+
+type LocationsNotFound struct {
+	image string
+}
+
+func (n LocationsNotFound) Error() string {
+	return fmt.Sprintf("Locations image in %s could not be found", n.image)
+}
+
+type Locations struct {
+	reader LocationImageReader
+	logger util.LoggerWithLevels
+}
+
+type LocationImageReader interface {
+	Read(img regv1.Image) (LocationConfig, error)
+}
+
+func NewLocations(logger util.LoggerWithLevels) *Locations {
+	return NewLocationsWithReader(&locationsSingleLayerReader{}, logger)
+}
+
+func NewLocationsWithReader(reader LocationImageReader, logger util.LoggerWithLevels) *Locations {
+	return &Locations{reader: reader, logger: logger}
+}
+
+func (r Locations) Fetch(registry image.ImagesMetadata, bundleRef name.Digest) (LocationConfig, error) {
+	r.logger.Tracef("fetching Locations OCI Images for bundle: %s\n", bundleRef)
+	locTag, err := r.tagFromBundleRef(bundleRef)
+	if err != nil {
+		return LocationConfig{}, fmt.Errorf("calculating locations image tag: %s", err)
+	}
+
+	img, err := registry.Image(locTag)
+	if err != nil {
+		if terr, ok := err.(*transport.Error); ok {
+			if terr.StatusCode == http.StatusNotFound {
+				return LocationConfig{}, &LocationsNotFound{image: locTag.Name()}
+			}
+		}
+		return LocationConfig{}, fmt.Errorf("fetching location image: %s", err)
+	}
+
+	r.logger.Tracef("reading the Locations configuration file\n")
+	cfg, err := r.reader.Read(img)
+	if err != nil {
+		return LocationConfig{}, fmt.Errorf("reading fetched location image: %s", err)
+	}
+
+	return cfg, err
+}
+
+func (r Locations) Save(reg ImagesMetadataWriter, bundleRef name.Digest, config LocationConfig, ui ui.UI) error {
+	r.logger.Tracef("saving Locations OCI Image for bundle: %s\n", bundleRef.Name())
+
+	tag, err := r.tagFromBundleRef(bundleRef)
+	if err != nil {
+		return fmt.Errorf("calculating locations image tag: %s", err)
+	}
+
+	tmpDir, err := os.MkdirTemp("", "locations")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmpDir)
```

Comment:
im pretty sure it's erroring because directory not empty. os.RemoveAll?

Reply:
fixed